pkg/kubernetes/client/v1alpha1/client.go

```diff
@@ -0,0 +1,165 @@
+//
+// NOTE:
+//  functional makes it simple to mock
+type GetConfigFunc func(*Client) (*rest.Config, error)
+
+// GetConfig returns kubernetes config instance
+//
+// NOTE:
+//  This is an implementation of GetConfigFunc
+func GetConfig(c *Client) (*rest.Config, error) {
+	if c == nil {
+		return nil, errors.New("failed to get kubernetes config: nil client was provided")
+	}
+	return c.Config()
+}
+
+// Client provides common kuberenetes client operations
+type Client struct {
+	IsInCluster            bool                       // flag to let client point to its own cluster
+	getInClusterConfig     getInClusterConfigFunc     // typed func to get in cluster config
+	buildConfigFromFlags   buildConfigFromFlagsFunc   // typed func to get desired kubernetes config
+	getKubernetesClientset getKubernetesClientsetFunc // typed func to get kubernetes clienset
+	getKubeMasterIP        getKubeMasterIPFunc        // typed func to get kubernetes master IP
+	getKubeConfigPath      getKubeConfigPathFunc      // typed func to get kubernetes config path
+}
+
+// OptionFunc is a typed function that abstracts any kind of operation
+// against the provided client instance
+//
+// This is the basic building block to create functional operations
+// against the client instance
+type OptionFunc func(*Client)
+
+// New returns a new instance of client
+func New(opts ...OptionFunc) *Client {
+	k := &Client{}
+	for _, o := range opts {
+		o(k)
+	}
+	if k.getInClusterConfig == nil {
+		k.getInClusterConfig = rest.InClusterConfig
+	}
+	if k.buildConfigFromFlags == nil {
+		k.buildConfigFromFlags = clientcmd.BuildConfigFromFlags
+	}
+	if k.getKubernetesClientset == nil {
+		k.getKubernetesClientset = kubernetes.NewForConfig
+	}
+	if k.getKubeMasterIP == nil {
+		k.getKubeMasterIP = env.Get
+	}
+	if k.getKubeConfigPath == nil {
+		k.getKubeConfigPath = env.Get
+	}
+	return k
+}
+
+// InCluster enables IsInCluster flag
+func InCluster() OptionFunc {
+	return func(c *Client) {
+		c.IsInCluster = true
+	}
+}
+
+// config returns the kubernetes config instance based on available criteria
+func (c *Client) Config() (config *rest.Config, err error) {
+	if c.IsInCluster {
+		return c.getInClusterConfig()
+	}
+	if strings.TrimSpace(c.getKubeMasterIP(K8sMasterIPEnvironmentKey)) != "" ||
+		strings.TrimSpace(c.getKubeConfigPath(KubeConfigEnvironmentKey)) != "" {
+		return c.getConfigFromENV()
+	}
+	return c.getInClusterConfig()
+}
+
+func (c *Client) getConfigFromENV() (config *rest.Config, err error) {
+	k8sMaster := c.getKubeMasterIP(K8sMasterIPEnvironmentKey)
+	kubeConfig := c.getKubeConfigPath(KubeConfigEnvironmentKey)
+	if strings.TrimSpace(k8sMaster) == "" &&
+		strings.TrimSpace(kubeConfig) == "" {
+		return nil, errors.Errorf("failed to get kubernetes config: missing environment variables: atleast one should be set: %s or %s", K8sMasterIPEnvironmentKey, KubeConfigEnvironmentKey)
+	}
+	return c.buildConfigFromFlags(k8sMaster, kubeConfig)
```

Comment:
We have handled the empty values for `k8sMaster` and `kubeConfig` (line no 151) but it is already checked before calling this method. That means there could be multiple callers --may be our intention is that we want this function(`getConfigFromENV` function) to be used by different callers and error handling should be taken care here in the `getConfigFromENV` function. Sounds good.
Now -- 1. If we want this function to called by different callers then on this principle( stated above) `c.buildConfigFromFlags` should be also checked for its empty value or for both cases it should bot be checked.

Reply:
validations are present in each method as they might be called by different callers.